// @temporalio/client/lib/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeQueryRejectCondition = exports.encodeQueryRejectCondition = exports.QueryRejectCondition = exports.InternalConnectionLikeSymbol = exports.HealthService = exports.TestService = exports.OperatorService = exports.WorkflowService = void 0;
const internal_workflow_1 = require("@temporalio/common/lib/internal-workflow");
const proto = __importStar(require("@temporalio/proto"));
exports.WorkflowService = proto.temporal.api.workflowservice.v1.WorkflowService;
exports.OperatorService = proto.temporal.api.operatorservice.v1.OperatorService;
exports.TestService = proto.temporal.api.testservice.v1.TestService;
exports.HealthService = proto.grpc.health.v1.Health;
exports.InternalConnectionLikeSymbol = Symbol('__temporal_internal_connection_like');
exports.QueryRejectCondition = {
    NONE: 'NONE',
    NOT_OPEN: 'NOT_OPEN',
    NOT_COMPLETED_CLEANLY: 'NOT_COMPLETED_CLEANLY',
    /** @deprecated Use {@link NONE} instead. */
    QUERY_REJECT_CONDITION_NONE: 'NONE', // eslint-disable-line deprecation/deprecation
    /** @deprecated Use {@link NOT_OPEN} instead. */
    QUERY_REJECT_CONDITION_NOT_OPEN: 'NOT_OPEN', // eslint-disable-line deprecation/deprecation
    /** @deprecated Use {@link NOT_COMPLETED_CLEANLY} instead. */
    QUERY_REJECT_CONDITION_NOT_COMPLETED_CLEANLY: 'NOT_COMPLETED_CLEANLY', // eslint-disable-line deprecation/deprecation
    /** @deprecated Use `undefined` instead. */
    QUERY_REJECT_CONDITION_UNSPECIFIED: undefined, // eslint-disable-line deprecation/deprecation
};
_a = (0, internal_workflow_1.makeProtoEnumConverters)({
    [exports.QueryRejectCondition.NONE]: 1,
    [exports.QueryRejectCondition.NOT_OPEN]: 2,
    [exports.QueryRejectCondition.NOT_COMPLETED_CLEANLY]: 3,
    UNSPECIFIED: 0,
}, 'QUERY_REJECT_CONDITION_'), exports.encodeQueryRejectCondition = _a[0], exports.decodeQueryRejectCondition = _a[1];
