// @temporalio/worker/lib/workflow/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initLoggerSink = initLoggerSink;
const common_1 = require("@temporalio/common");
const logger_1 = require("@temporalio/common/lib/logger");
/**
 * Injects a logger sink that forwards to the worker's logger
 */
function initLoggerSink(logger) {
    logger = logger_1.LoggerWithComposedMetadata.compose(logger, { sdkComponent: common_1.SdkComponent.workflow });
    return {
        __temporal_logger: {
            trace: {
                fn(_, message, attrs) {
                    logger.trace(message, attrs);
                },
            },
            debug: {
                fn(_, message, attrs) {
                    logger.debug(message, attrs);
                },
            },
            info: {
                fn(_, message, attrs) {
                    logger.info(message, attrs);
                },
            },
            warn: {
                fn(_, message, attrs) {
                    logger.warn(message, attrs);
                },
            },
            error: {
                fn(_, message, attrs) {
                    logger.error(message, attrs);
                },
            },
        },
    };
}
