// @temporalio/common/lib/index.js
"use strict";
/**
 * Common library for code that's used across the Client, Worker, and/or Workflow
 *
 * @module
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineSearchAttributeKey = exports.TypedSearchAttributes = exports.SearchAttributeType = exports.ActivityCancellationDetails = void 0;
exports.u8 = u8;
exports.str = str;
exports.errorMessage = errorMessage;
exports.errorCode = errorCode;
const encoding = __importStar(require("./encoding"));
const helpers = __importStar(require("./type-helpers"));
__exportStar(require("./activity-options"), exports);
var activity_cancellation_details_1 = require("./activity-cancellation-details");
Object.defineProperty(exports, "ActivityCancellationDetails", { enumerable: true, get: function () { return activity_cancellation_details_1.ActivityCancellationDetails; } });
__exportStar(require("./converter/data-converter"), exports);
__exportStar(require("./converter/failure-converter"), exports);
__exportStar(require("./converter/payload-codec"), exports);
__exportStar(require("./converter/payload-converter"), exports);
__exportStar(require("./converter/types"), exports);
__exportStar(require("./deprecated-time"), exports);
__exportStar(require("./errors"), exports);
__exportStar(require("./failure"), exports);
__exportStar(require("./interfaces"), exports);
__exportStar(require("./logger"), exports);
__exportStar(require("./priority"), exports);
__exportStar(require("./metrics"), exports);
__exportStar(require("./retry-policy"), exports);
__exportStar(require("./worker-deployments"), exports);
__exportStar(require("./workflow-definition-options"), exports);
__exportStar(require("./workflow-handle"), exports);
__exportStar(require("./workflow-options"), exports);
__exportStar(require("./versioning-intent"), exports);
var search_attributes_1 = require("./search-attributes");
Object.defineProperty(exports, "SearchAttributeType", { enumerable: true, get: function () { return search_attributes_1.SearchAttributeType; } });
Object.defineProperty(exports, "TypedSearchAttributes", { enumerable: true, get: function () { return search_attributes_1.TypedSearchAttributes; } });
Object.defineProperty(exports, "defineSearchAttributeKey", { enumerable: true, get: function () { return search_attributes_1.defineSearchAttributeKey; } });
/**
 * Encode a UTF-8 string into a Uint8Array
 *
 * @hidden
 * @deprecated - meant for internal use only
 */
function u8(s) {
    return encoding.encode(s);
}
/**
 * Decode a Uint8Array into a UTF-8 string
 *
 * @hidden
 * @deprecated - meant for internal use only
 */
function str(arr) {
    return encoding.decode(arr);
}
/**
 * Get `error.message` (or `undefined` if not present)
 *
 * @hidden
 * @deprecated - meant for internal use only
 */
function errorMessage(error) {
    return helpers.errorMessage(error);
}
/**
 * Get `error.code` (or `undefined` if not present)
 *
 * @hidden
 * @deprecated - meant for internal use only
 */
function errorCode(error) {
    return helpers.errorCode(error);
}
